### PR TITLE
feat(events): track supply changes more closely PE-6796

### PR DESCRIPTION
**Pull request by arielmelendez**

### CODE REVIEW

File: src/main.lua [hunk 1]
```diff
@@ -809,7 +870,7 @@ addEventingHandler(ActionMap.JoinNetwork, utils.hasMatchingTag("Action", ActionM
 	local timestamp = tonumber(msg.Timestamp)
 
 	msg.ioEvent:addField("Resolved-Observer-Address", formattedObserverAddress)
-	msg.ioEvent:addField("Sender-Previous-Balance", fromBalance)
+	msg.ioEvent:addField("Sender-Previous-Balance", Balances[fromAddress] or 0)
```
Comment (dtfiedler) [inline src/main.lua:873]: nit - 
```suggestion
	msg.ioEvent:addField("Sender-Previous-Balance", balances.getBalance(fromAddress))
```

File: src/main.lua [hunk 2]
```diff
@@ -822,14 +883,19 @@ addEventingHandler(ActionMap.JoinNetwork, utils.hasMatchingTag("Action", ActionM
 		return
 	end
 
-	msg.ioEvent:addField("Sender-New-Balance", fromBalance)
+	msg.ioEvent:addField("Sender-New-Balance", Balances[fromAddress] or 0)
```
Comment (dtfiedler) [inline src/main.lua:886]:
```suggestion
	msg.ioEvent:addField("Sender-New-Balance", balances.getBalance(fromAddress))
```

File: src/vaults.lua [hunk 3]
```diff
@@ -139,18 +139,22 @@ end
...
 				balances.increaseBalance(owner, nestedVault.balance)
-				vaults[id] = nil
+				ownersVaults[id] = nil
+				prunedVaults[id] = nestedVault
 			end
 		end
 		-- update the owner vault
-		allVaults[owner] = vaults
+		-- TODO: I THINK THIS LINE IS UNNECESSARY. CHECK TESTS
+		allVaults[owner] = ownersVaults
```
Comment (dtfiedler) [inline src/vaults.lua:153]: it's not - as we are updating the objects in place above, so not necessary - this was likely leftover from when we ported it over from smartweave

File: src/main.lua (hunk 1 above)
Comment (arielmelendez) [reply src/main.lua:873]: My issue with that approach is it makes a whole deep copy of the balances table to then just throw it away. Why do that here for a scalar value?

File: src/vaults.lua (hunk 3 above)
Comment (arielmelendez) [reply src/vaults.lua]: Excised and tests ran fine:
be82619

File: src/main.lua (hunk 1 above)
Comment (arielmelendez) [reply src/main.lua:873]: Perhaps what we need is a cheaper getter function if you want to protect accesses of the main Balances object?

File: src/main.lua [hunk 4]
```diff
@@ -882,6 +955,11 @@ addEventingHandler(ActionMap.LeaveNetwork, utils.hasMatchingTag("Action", Action
 	msg.ioEvent:addField("Joined-Gateways-Count", gwStats.joined)
 	msg.ioEvent:addField("Leaving-Gateways-Count", gwStats.leaving)
 
+	lastKnownStakedSupply = lastKnownStakedSupply - gwPrevStake - gwPrevTotalDelegatedStake
+	lastKnownWithdrawSupply = lastKnownWithdrawSupply + gwPrevStake + gwPrevTotalDelegatedStake
+	msg.ioEvent:addField("Staked-Supply", lastKnownStakedSupply)
+	msg.ioEvent:addField("Withdraw-Supply", lastKnownWithdrawSupply)
```
Comment (arielmelendez) [inline src/main.lua:961]: TODO: Add event payload data for gw's remaining totalDelegatedStake which should damn well be 0!

File: src/gar.lua [hunk 5]
```diff
@@ -654,6 +655,9 @@ function gar.pruneGateways(currentTimestamp, msgId)
 	local result = {
 		prunedGateways = {},
 		slashedGateways = {},
+		gatewayStakeReturned = 0,
+		delegateStakeReturned = 0,
+		stakeSlashed = 0,
```
Comment (arielmelendez) [inline src/gar.lua:660]: TODO: Add pruned gws with non-zero totalDelegatedStake to use in event data.

File: src/main.lua [hunk 6]
```diff
@@ -197,28 +197,60 @@ end, function(msg)
...
 		end
 	end
 
+	msg.ioEvent:addField(
+		"Total-Supply",
+		lastKnownCirculatingSupply
+			+ lastKnownLockedSupply
+			+ lastKnownStakedSupply
+			+ lastKnownDelegatedSupply
+			+ lastKnownWithdrawSupply
```
Comment (arielmelendez) [inline src/main.lua:245]: Todo: add protocol balance

File: src/main.lua [hunk 7]
```diff
@@ -197,28 +197,60 @@ end, function(msg)
...
+		local slashedGatewaysCount = utils.lengthOfTable(slashedGateways or {})
 		if slashedGatewaysCount > 0 then
-			msg.ioEvent:addField("Slashed-Gateways", resultOrError.slashedGateways)
+			msg.ioEvent:addField("Slashed-Gateways", slashedGateways)
 			msg.ioEvent:addField("Slashed-Gateways-Count", slashedGatewaysCount)
 		end
 	end
 
+	msg.ioEvent:addField(
+		"Total-Supply",
```
Comment (arielmelendez) [inline src/main.lua:240]: Update to Total-Token-Supply

File: src/main.lua [hunk 8]
```diff
@@ -197,28 +197,60 @@ end, function(msg)
...
+		lastKnownCirculatingSupply
+			+ lastKnownLockedSupply
+			+ lastKnownStakedSupply
+			+ lastKnownDelegatedSupply
+			+ lastKnownWithdrawSupply
+	)
+	msg.ioEvent:addField("Circulating-Supply", lastKnownCirculatingSupply)
```
Comment (arielmelendez) [inline src/main.lua:247]: Consider only setting these here if there is a change in value.

File: src/main.lua (hunk 7 above)
Comment (arielmelendez) [reply src/main.lua]: 6b1688e

File: src/main.lua (hunk 6 above)
Comment (arielmelendez) [reply src/main.lua]: 6b1688e

File: src/main.lua (hunk 8 above)
Comment (arielmelendez) [reply src/main.lua]: 6b1688e

File: src/main.lua (hunk 4 above)
Comment (arielmelendez) [reply src/main.lua:990]: cfec28d

File: src/gar.lua (hunk 5 above)
Comment (arielmelendez) [reply src/gar.lua:660]: cfec28d

**Pull request merged**